spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,201 @@
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new PrunedFilteredScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and Catalyst filtered table scan relation*/
+case object CatalystScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new CatalystScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Table scan type */
+sealed trait ScanType {
+  def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                   userSpecifiedSchema: Option[StructType], sqlContext: SQLContext) : BaseRelationImpl
+}
+
+/** Base relation implements [[BaseRelation]] and [[InsertableRelation]] */
+private[cassandra] class BaseRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                          userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelation with InsertableRelation with Serializable with Logging {
+
+  override def sqlContext: SQLContext = sqlcontext
+
+  protected[this] val tableDef = sqlContext
+    .schemas.get(cluster.getOrElse("default"))
+    .keyspaceByName.getOrElse(keyspace, throw new IOException(s"Keyspace not found: $keyspace"))
+    .tableByName.getOrElse(table, throw new IOException(s"Table not found: $keyspace.$table"))
+
+  protected[this] val columnNameByLowercase = tableDef.allColumns
+    .map(c => (c.columnName.toLowerCase, c.columnName)).toMap
+
+  override def schema: StructType = {
+    def columnToStructField(column: ColumnDef): StructField = {
+      StructField(column.columnName, ColumnDataType.catalystDataType(column.columnType, nullable = true))
+    }
+    if (userSpecifiedSchema.isDefined){
+      userSpecifiedSchema.get
+    } else {
+      StructType(tableDef.allColumns.map(columnToStructField))
+    }
+  }
+
+  protected[this] val baseRdd = sqlContext
+    .sparkContext
+    .cassandraTable[CassandraSQLRow](keyspace, table)(
+      new CassandraConnector(sqlContext.getCassandraConnConf(cluster)),
+      sqlContext.getReadConf(keyspace, table, cluster),
+      implicitly[ClassTag[CassandraSQLRow]],
+      CassandraSQLRowReader,
+      implicitly[ValidRDDType[CassandraSQLRow]])
+
+  private[this] val writeConf = sqlContext.getWriteConf(keyspace, table, cluster)
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    data.rdd.saveToCassandra(keyspace, table, AllColumns, writeConf)(
+        new CassandraConnector(sqlContext.getCassandraConnConf(cluster)), SqlRowWriter.Factory)
+  }
+
+  def buildScan(): RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
+
+  protected[this] def getSchemaData(column: String, row: Row): (Any, NativeType) = {
```

Comment:
This method seems to be called very frequently and thus its implementation may affect performance. Consider precomputing a map `columnName -> (index, columnDef, dataType)`.